python/pyarrow/tensorflow/plasma_op.cc

```diff
@@ -0,0 +1,381 @@
+      CHECK_GT(s, 0);
+      total_bytes += s;
+      offsets.push_back(total_bytes);
+    }
+
+    const Tensor& plasma_object_id = context->input(num_inputs - 1);
+    CHECK_EQ(plasma_object_id.NumElements(), 1);
+    const string& plasma_object_id_str = plasma_object_id.flat<std::string>()(0);
+    VLOG(1) << "plasma_object_id_str: '" << plasma_object_id_str << "'";
+    const plasma::ObjectID object_id =
+        plasma::ObjectID::from_binary(plasma_object_id_str);
+
+    std::vector<int64_t> shape = {total_bytes / byte_width};
+
+    arrow::io::MockOutputStream mock;
+    ARROW_CHECK_OK(arrow::py::WriteTensorHeader(arrow_dtype, shape, 0, &mock));
+    int64_t header_size = mock.GetExtentBytesWritten();
+
+    std::shared_ptr<Buffer> data_buffer;
+    {
+      mutex_lock lock(mu_);
+      ARROW_CHECK_OK(client_.Create(object_id, header_size + total_bytes,
+                                    /*metadata=*/nullptr, 0, &data_buffer));
+    }
+
+    int64_t offset;
+    arrow::io::FixedSizeBufferWriter buf(data_buffer);
+    ARROW_CHECK_OK(arrow::py::WriteTensorHeader(arrow_dtype, shape, total_bytes, &buf));
+    ARROW_CHECK_OK(buf.Tell(&offset));
+
+    uint8_t* data = reinterpret_cast<uint8_t*>(data_buffer->mutable_data() + offset);
+
+    auto wrapped_callback = [this, context, done, data_buffer, object_id]() {
+      {
+        mutex_lock lock(mu_);
+        ARROW_CHECK_OK(client_.Seal(object_id));
+      }
+      context->SetStatus(tensorflow::Status::OK());
+      done();
+    };
+
+    if (std::is_same<Device, CPUDevice>::value) {
+      for (int i = 0; i < num_tensors; ++i) {
+        const auto& input_tensor = context->input(i);
+        std::memcpy(static_cast<void*>(data + offsets[i]),
+                    input_tensor.tensor_data().data(),
+                    static_cast<uint64>(offsets[i + 1] - offsets[i]));
+      }
+      wrapped_callback();
+    } else {
+#ifdef GOOGLE_CUDA
+      auto orig_stream = context->op_device_context()->stream();
+      OP_REQUIRES_ASYNC(context, orig_stream != nullptr,
+                        errors::Internal("No GPU stream available."), done);
+      auto stream_executor = orig_stream->parent();
+
+      // NOTE(zongheng): this is critical of getting good performance out of D2H
+      // async memcpy.  Under the hood it performs cuMemHostRegister(), see:
+      // http://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gf0a9fe11544326dabd743b7aa6b54223
+      CHECK(stream_executor->HostMemoryRegister(static_cast<void*>(data),
+                                                static_cast<uint64>(total_bytes)));
+
+      {
+        mutex_lock l(d2h_stream_mu);
+        if (d2h_stream == nullptr) {
+          d2h_stream = new Stream(stream_executor);
+          CHECK(d2h_stream->Init().ok());
+        }
+      }
+
+      // Needed to make sure the input buffers have been computed.
+      // NOTE(ekl): this is unnecessary when the op is behind a NCCL allreduce already
```

Comment:
who is ekl?

Reply:
ekl is @ericl, also a student at the RISELab who helped getting this together